ts/query.go

```diff
@@ -0,0 +1,411 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Matt Tracy (matt@cockroachlabs.com)
+
+package ts
+
+import (
+	"container/heap"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// dataSpan is used to construct a normalized, monolithic view of a single time
+// series over an arbitrary time span. The actual data in a span may be stored
+// in multiple instances of InternalTimeSeriesData.
+type dataSpan struct {
+	startNanos  int64
+	sampleNanos int64
+	datas       []normalizedData
+}
+
+// addData adds an InternalTimeSeriesData object into this dataSpan, normalizing
+// it to a normalizedData object in the process.
+func (ds *dataSpan) addData(data *proto.InternalTimeSeriesData) error {
+	if data.SampleDurationNanos != ds.sampleNanos {
+		return util.Errorf("data added to dataSpan with mismatched sample duration period.")
+	}
+
+	// Calculate an offset adjustment which normalizes the supplied data into
+	// the dataSpan's time period.
+	adjustment := (data.StartTimestampNanos - ds.startNanos) / ds.sampleNanos
+	rd := normalizedData{data, int32(adjustment)}
+
+	// If all samples in the supplied data set are before normalized offset 0,
+	// do not include it.
+	if rd.offsetAt(len(data.Samples)-1) < 0 {
+		return nil
+	}
+
+	// If the supplied data does not occur after all previously added data,
+	// return an error.
+	if len(ds.datas) > 0 {
+		last := ds.datas[len(ds.datas)-1]
+		if rd.offsetAt(0) <= last.offsetAt(len(last.Samples)-1) {
+			return util.Error("data must be added to dataSpan in chronological order.")
+		}
+	}
+
+	ds.datas = append(ds.datas, rd)
+	return nil
+}
+
+// timestampForOffset returns an appropriate timestamp for the given offset
+// within a dataSpan. Because each offset represents a duration of time and not
+// an exact time, the returned timestamp will fall exactly in the middle of the
+// time slot represented by the offset.
+func (ds *dataSpan) timestampForOffset(offset int32) int64 {
+	return ds.startNanos + (int64(offset) * ds.sampleNanos) + (ds.sampleNanos / 2)
+}
+
+// normalizedData is used to "normalize" an InternalTimeSeriesData object for
+// use in a dataSpan.  This is accomplished by computing a constant offset
+// adjustment which adjusts each Sample offset to be relative to the start time
+// of the dataSpan, rather than the start time of the InternalTimeSeriesData.
+type normalizedData struct {
+	*proto.InternalTimeSeriesData
+	offsetAdjustment int32
+}
+
+// offsetAt returns the normalized offset for the Sample at the supplied index
+// in the InternalTimeSeriesData's Samples collection.
+func (rdc *normalizedData) offsetAt(idx int) int32 {
```

Comment:
I may have the wrong mental model, but in the case of your testing with Resolution=1ns, wouldn't the int32 here be a practical problem in any event except unittesting with base time close to 0?